### PR TITLE
Return images same mode (initial commit)

### PR DESCRIPTION
**Pull request by membriux**
## Related Issue
Fixes #{128}

## Summary
- [x] I have read CONTRIBUTING.md to understand how to contribute to this repository :)

- [x] Add src_mode arg to ret_and_save_image() in image/utils/utils.py
- [x] Pass src_mode arg into ret_and_save_image() from every augmentation except convert_color in image/functional.py (e.g. here for apply_lambda)
- [x] In image test evaluate_class(), assert that the mode of self.img & dst are equal.
- [x] Run image tests, make sure they all pass: python -m unittest discover -s augly/tests/image_tests/ -p "*"

## Unit Tests
If your changes touch the `audio` module, please run all of the `audio` tests and paste the output here. Likewise for `image`, `text`, & `video`. If your changes could affect behavior in multiple modules, please run the tests for all potentially affected modules. If you are unsure of which modules might be affected by your changes, please just run all the unit tests.

### [Image](url)

```bash
python -m unittest discover -s augly/tests/image_tests/ -p "*_test.py"
# Or `python -m unittest discover -s augly/tests/image_tests/ -p "*.py"` to run pytorch test too (must install `torchvision` to run)
```

## Other testing

N/A

### CODE REVIEW

File: augly/image/utils/utils.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def validate_and_load_image(image: Union[str, Image.Image]) -> Image.Image:
     return image
 
 
-def ret_and_save_image(image: Image.Image, output_path: Optional[str]) -> Image.Image:
+def ret_and_save_image(image: Image.Image, output_path: Optional[str], src_mode: Optional[Image.Image]) -> Image.Image:
```
Comment (zpapakipos) [inline augly/image/utils/utils.py:33]:
```suggestion
def ret_and_save_image(
    image: Image.Image, output_path: Optional[str], src_mode: str
) -> Image.Image:
```

Make sure no line has >90 columns; also `src_mode` should be a string -- e.g. "RGBA", "RGB", or "L" :)

Comment (zpapakipos) [reply augly/image/utils/utils.py:33]: Also, now that you've added `src_mode` as a param to this function, you should use it! For example you could change line 41 to:
```python4
return image.convert(src_mode)
```

File: augly/image/functional.py [hunk 2]
```diff
@@ -74,7 +74,7 @@ def apply_lambda(
         **func_kwargs,
     )
 
-    return imutils.ret_and_save_image(aug_image, output_path)
+    return imutils.ret_and_save_image(aug_image, output_path, src_mode)
```
Comment (zpapakipos) [inline augly/image/functional.py:77]: This looks like what we want to do, but you haven't defined `src_mode` yet. You can get the mode of the src image like this (be sure to add this line in each image function **before** `image` is augmented in any way but **after** `validate_and_load_image()` is called & **after** `get_func_kwargs()` is called (so `src_mode` isn't added to the metadata).
```python
src_mode = image.mode
```
(Source: https://stackoverflow.com/questions/33831572/get-image-mode-pil-python)

File: augly/tests/image_tests/base_unit_test.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ def evaluate_class(
         dst = transform_class(
             self.img, metadata=metadata, bboxes=bboxes, bbox_format=bbox_format
         )
-
+        self.assertTrue(are_equal_images(self.img, dst))
```
Comment (zpapakipos) [inline augly/tests/image_tests/base_unit_test.py:106]: Be careful here -- we actually want to check that the **mode** of `self.img` & `dst` are equal, not that the entire images are equal. For example for `test_Crop` this would fail, because `dst` is the cropped image, which is smaller than `self.img` (the input image).
```suggestion
        self.assertTrue(self.img.mode == dst.mode)
```

**Review by zpapakipos — CHANGES_REQUESTED** [review]
Thanks for this PR, @membriux! Overall you're definitely on the right track & you've identified most of the places where changes need to be made, but you're still missing a few steps. Please address my comments **and** make sure to actually run the image tests locally (and paste the output in the initial PR comment please, so I can see they're passing before running the GitHub test workflows).

File: augly/image/utils/utils.py (hunk 1 above)
Comment (membriux) [reply augly/image/utils/utils.py:33]: Will the `src_mode` be optional?

File: augly/tests/image_tests/base_unit_test.py (hunk 3 above)
Comment (membriux) [reply augly/tests/image_tests/base_unit_test.py:106]: @zpapakipos how would this work for `convert_color()` since we are excluding it?

**Comment by membriux** [comment]
@zpapakipos I added the changes. How should I go about `convertColor` and the Class-based functions (see output of Test cases below)?

```
======================================================================
ERROR: test_convert_color (functional_unit_test.FunctionalImageUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/functional_unit_test.py", line 37, in test_convert_color
    self.evaluate_function(imaugs.convert_color, mode="L")
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/base_unit_test.py", line 87, in evaluate_function
    aug_function(self.local_img_path, output_path=tmpfile.name, **kwargs)
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/image/functional.py", line 561, in convert_color
    return imutils.ret_and_save_image(aug_image, output_path)
TypeError: ret_and_save_image() missing 1 required positional argument: 'src_mode'

======================================================================
ERROR: test_ConvertColor (transforms_unit_test.TransformsImageUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/transforms_unit_test.py", line 70, in test_ConvertColor
    self.evaluate_class(imaugs.ConvertColor(mode="L"), fname="convert_color")
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/base_unit_test.py", line 103, in evaluate_class
    dst = transform_class(
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/image/transforms.py", line 60, in __call__
    return self.apply_transform(image, metadata, bboxes, bbox_format)
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/image/transforms.py", line 646, in apply_transform
    return F.convert_color(
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/image/functional.py", line 561, in convert_color
    return imutils.ret_and_save_image(aug_image, output_path)
TypeError: ret_and_save_image() missing 1 required positional argument: 'src_mode'

======================================================================
FAIL: test_Compose (transforms_unit_test.TransformsImageUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/transforms_unit_test.py", line 49, in test_Compose
    self.evaluate_class(
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/base_unit_test.py", line 106, in evaluate_class
    self.assertTrue(self.img.mode == dst.mode)
AssertionError: False is not true

======================================================================
FAIL: test_Opacity (transforms_unit_test.TransformsImageUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/transforms_unit_test.py", line 100, in test_Opacity
    self.evaluate_class(imaugs.Opacity(), fname="opacity")
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/base_unit_test.py", line 106, in evaluate_class
    self.assertTrue(self.img.mode == dst.mode)
AssertionError: False is not true

======================================================================
FAIL: test_OverlayEmoji (transforms_unit_test.TransformsImageUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/transforms_unit_test.py", line 103, in test_OverlayEmoji
    self.evaluate_class(imaugs.OverlayEmoji(), fname="overlay_emoji")
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/base_unit_test.py", line 106, in evaluate_class
    self.assertTrue(self.img.mode == dst.mode)
AssertionError: False is not true

======================================================================
FAIL: test_OverlayImage (transforms_unit_test.TransformsImageUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/transforms_unit_test.py", line 106, in test_OverlayImage
    self.evaluate_class(
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/base_unit_test.py", line 106, in evaluate_class
    self.assertTrue(self.img.mode == dst.mode)
AssertionError: False is not true

======================================================================
FAIL: test_OverlayOntoBackgroundImage (transforms_unit_test.TransformsImageUnitTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/transforms_unit_test.py", line 112, in test_OverlayOntoBackgroundImage
    self.evaluate_class(
  File "/Users/macbookpro/Desktop/Github/AugLy/augly/tests/image_tests/base_unit_test.py", line 106, in evaluate_class
    self.assertTrue(self.img.mode == dst.mode)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 82 tests in 54.550s

FAILED (failures=5, errors=2, skipped=5)
```

**Review by zpapakipos — CHANGES_REQUESTED** [review]
I made comments in your code which should fix 2 of the 10 failing tests (`test_opacity` & `test_ConvertColor`).

For the 7 tests which are failing with `ValueError: operands could not be broadcast together with shapes (w,h,3) (w,h,4)` (`test_overlay_emoji`, `test_overlay_image`, `test_Compose`, `test_Opacity`, `test_OverlayEmoji`, `test_OverlayImage`, & `test_OverlayOntoBackgroundImage`), this is because you are now returning the image from all augmentations in the same mode as the input image was passed in, and for these 7 augmentations the image was previously being returned in a different mode (RGBA instead of RGB, which we can tell because the number of channels is 3 which doesn't match the expected 4). To fix this, you will need to replace the expected output image files for these 7 tests (they live in [this](https://github.com/facebookresearch/AugLy/tree/main/augly/assets/tests/image/dfdc_expected_output) directory). You can generate the files that are now created by these augmentations' unit tests by e.g. changing [this line](https://github.com/facebookresearch/AugLy/blob/main/augly/tests/image_tests/base_unit_test.py#L87) to `aug_function(self.local_img_path, output_path=f"/tmp/{aug_function.__name__}.png", **kwargs)`. Then run the 7 tests (e.g. `python -m unittest augly.tests.image_tests.functional_unit_test.test_overlay_emoji`) and the new output files will be at e.g. `/tmp/test_overlay_emoji.png` in your local file system, so you can copy them to where they need to go with e.g. `cp /tmp/test_overlay_emoji.png augly/assets/tests/image/dfdc_expected_output/test_overlay_emoji.png`.

For the final failing test (`test_overlay_onto_background_image`), **if** you still get this error after fixing the above tests, I'm not immediately sure what this error message (`ValueError: image has no palette`) means so you'll have to do a little debugging. I would recommend adding some print statements & then running just this unit test (`python -m unittest augly.tests.image_tests.functional_unit_test.FunctionalImageUnitTest.test_overlay_onto_background_image`) to try to see where something could be going wrong. Off the top of my head I suspect it might have to do with the background image we pass in in the test being in a weird mode, so I would definitely `print(background_image.mode)` ([here](https://github.com/facebookresearch/AugLy/blob/main/augly/image/functional.py#L1242)). You should also try googling the error message to see if others have pinpointed what the issue causing that generally is.

Let me know if you have any questions or get stuck on any of the above! :)

**Comment by membriux** [comment]
@zpapakipos When I attempted to generate the files created by the augmentations i got an error.

**What I did**
Changed line for output files on `base_unit_tests.py`: 
`aug_function(self.local_img_path, output_path=f"/tmp/{aug_function.__name__}.png", **kwargs)`

**Ran command**
`python -m unittest augly.tests.image_tests.functional_unit_test.test_overlay_emoji`

**Test Output*

```
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/__main__.py", line 12, in <module>
    main(module=None)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/main.py", line 94, in __init__
    self.parseArgs(argv)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/main.py", line 149, in parseArgs
    self.createTests()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/main.py", line 158, in createTests
    self.module)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/loader.py", line 130, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/loader.py", line 91, in loadTestsFromName
    module = __import__('.'.join(parts_copy))
  File "augly/tests/__init__.py", line 4, in <module>
    from augly.tests.base_configs import (
  File "augly/tests/base_configs.py", line 13
    output_dir: str
              ^
SyntaxError: invalid syntax
```

File: augly/image/functional.py [hunk 4]
```diff
@@ -992,8 +1011,9 @@ def opacity(
     assert 0 <= level <= 1, "level must be a value in the range [0, 1]"
 
     image = imutils.validate_and_load_image(image).convert(mode="RGBA")
-
+
```
Comment (zpapakipos) [inline augly/image/functional.py:1014]:
```suggestion

```

File: augly/image/functional.py [hunk 5]
```diff
@@ -1692,6 +1722,7 @@ def pad_square(
     image = imutils.validate_and_load_image(image)
 
     func_kwargs = imutils.get_func_kwargs(metadata, locals())
+    src_mode = image.mode
```
[the remark this answers is not in the document]
Comment (zpapakipos) [reply augly/image/functional.py:1722]: Unresolving because I still don't see the change in here! It should be on what is now line 1744.

File: augly/image/transforms.py [hunk 6]
```diff
@@ -1234,6 +1234,7 @@ def apply_transform(
 
         @returns: Augmented PIL Image
         """
+
```
Comment (zpapakipos) [inline augly/image/transforms.py:1237]:
```suggestion
```

File: augly/image/functional.py [hunk 7]
```diff
@@ -1058,6 +1078,7 @@ def overlay_emoji(
...
```
Comment (zpapakipos) [inline augly/image/functional.py:1081]: Same issue here as below! `src_mode` isn't getting passed into `ret_and_save_image()` because it isn't called in this augmentation. That's a mistake, so please change line 1102 below to `return imutils.ret_and_save_image(aug_image, output_path, src_mode)`

**Review by zpapakipos — CHANGES_REQUESTED** [review]

File: augly/image/functional.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (membriux) [reply augly/image/functional.py:1722]: Pushed the fix for this one now :)

**Comment by zpapakipos** [comment]
Hi @membriux thanks for the detailed steps & output! A few things in response:

> Step 2
This is where we get some errors running this line on the terminal
python3 -m unittest augly.tests.image_tests.functional_unit_test.test_overlay_emoji

FYI, the issue here is that it should actually be `python3 -m unittest augly.tests.image_tests.functional_unit_test. FunctionalImageUnitTest.test_overlay_emoji`. You need the class name that `test_overlay_emoji` lives in in order for `unittest` to find it in that file :)

> Step 4
Run tests with new copied expected output images.

Are you sure you ran these tests from this branch? The error message you're getting on `self.assertTrue(self.img.mode == dst.mode)` isn't due to anything about the expected output image file, but is just checking that the mode is the same between the src image & the augmented image, i.e. the change you've made in this diff. And as I remember this line was passing before, so either you made a change which makes the mode changing no longer work, or perhaps you were running these tests on `main` instead of on this branch.

Also, can you please push all your changes to this PR? I don't see the new expected output image files in the "Files changed" section. Seeing all of your current changes which you're running would be very helpful for me in helping you debug, so I can look for obvious issues or even check out your branch & run the tests myself.

File: augly/image/functional.py [hunk 8]
```diff
@@ -58,12 +58,14 @@ def apply_lambda(
     image = imutils.validate_and_load_image(image)
 
     func_kwargs = deepcopy(locals())
+
```
Comment (zpapakipos) [inline augly/image/functional.py:61]:
```suggestion
```

File: augly/image/functional.py [hunk 9]
```diff
@@ -824,19 +837,21 @@ def masked_composite(
...
```
Comment (zpapakipos) [inline augly/image/functional.py:840]:
```suggestion
```

File: augly/image/utils/utils.py [hunk 10]
```diff
@@ -30,7 +30,13 @@ def validate_and_load_image(image: Union[str, Image.Image]) -> Image.Image:
     return image
 
 
-def ret_and_save_image(image: Image.Image, output_path: Optional[str]) -> Image.Image:
+def ret_and_save_image(
+    image: Image.Image, output_path: Optional[str], src_mode: Optional[str] = None
+) -> Image.Image:
+
```
Comment (zpapakipos) [inline augly/image/utils/utils.py:36]:
```suggestion
```

File: augly/tests/image_tests/base_unit_test.py [hunk 11]
```diff
@@ -63,6 +63,7 @@ def are_equal_metadata(
 
 class BaseImageUnitTest(unittest.TestCase):
     ref_img_dir = os.path.join(TEST_URI, "image", "dfdc_expected_output")
+
```
Comment (zpapakipos) [inline augly/tests/image_tests/base_unit_test.py:66]:
```suggestion
```

File: augly/tests/image_tests/base_unit_test.py [hunk 12]
```diff
@@ -96,13 +97,18 @@ def evaluate_class(
         transform_class: Callable[..., Image.Image],
         fname: str,
         metadata_exclude_keys: Optional[List[str]] = None,
+        check_mode: Optional[bool] = True
     ):
+
```
Comment (zpapakipos) [inline augly/tests/image_tests/base_unit_test.py:102]:
```suggestion
```

File: augly/tests/image_tests/transforms_unit_test.py [hunk 13]
```diff
@@ -67,7 +67,11 @@ def test_Contrast(self):
         self.evaluate_class(imaugs.Contrast(), fname="contrast")
 
     def test_ConvertColor(self):
-        self.evaluate_class(imaugs.ConvertColor(mode="L"), fname="convert_color")
+        self.evaluate_class(
+            imaugs.ConvertColor(mode="L"), 
+            fname="convert_color", 
+            check_mode=False
```
Comment (zpapakipos) [inline augly/tests/image_tests/transforms_unit_test.py:73]:
```suggestion
            check_mode=False,
```

**Comment by membriux** [comment]
Fixed up :)

File: augly/tests/image_tests/base_unit_test.py [hunk 14]
```diff
@@ -96,13 +96,17 @@ def evaluate_class(
         transform_class: Callable[..., Image.Image],
         fname: str,
         metadata_exclude_keys: Optional[List[str]] = None,
+        check_mode: Optional[bool] = True
```
Comment (jbitton) [inline augly/tests/image_tests/base_unit_test.py:99]:
```suggestion
        check_mode: Optional[bool] = True,
```
missing trailing comma

File: augly/image/functional.py [hunk 15]
```diff
@@ -64,6 +64,7 @@ def apply_lambda(
         except AttributeError:
             func_kwargs["aug_function"] = type(aug_function).__name__
     func_kwargs = imutils.get_func_kwargs(metadata, func_kwargs)
+    src_mode = image.mode
```
Comment (jbitton) [inline augly/image/functional.py:67]:
```suggestion
            func_kwargs["aug_function"] = type(aug_function).__name__

    func_kwargs = imutils.get_func_kwargs(metadata, func_kwargs)
    src_mode = image.mode
```
(not caused by your diff but i think this would be more readable)

File: augly/image/functional.py [hunk 16]
```diff
@@ -911,6 +924,7 @@ def meme_format(
 
     image = imutils.validate_and_load_image(image)
     func_kwargs = imutils.get_func_kwargs(metadata, locals())
+    src_mode = image.mode
```
Comment (jbitton) [inline augly/image/functional.py:927]: i think you forgot to use  this `src_mode` on line 969

File: augly/image/functional.py (hunk 7 above)
Comment (jbitton) [reply augly/image/functional.py:1076]: wait, @zpapakipos -- i don't think it's needed to have these extra calls to `ret_and_save_image`. this function, `overlay_emoji` calls within it `overlay_image` to do the actual overlaying of the emoji onto the image. thus, as long as `overlay_image` handles the modes correctly, *we do not need to add these extra function calls*. 

tl;dr you can remove `src_mode = image.mode` and the ret and save call from this function as they are already handled by `overlay_image`. -- let me know if you see what i mean @zpapakipos 

sorry for the extra work! if you've done this anywhere else, please undo those changes too (ill try to comment where i find them)

File: augly/image/functional.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbitton) [reply augly/image/functional.py:1717]: +1 on my previous comment -- you can undo these changes because as long as `pad` handles the modes correctly doing this in `pad_square` is redundant (cc: @zpapakipos)

**Review by jbitton — APPROVED** [review]
Overall LGTM! Great job on your first task!! A couple of final nits and then I think we should be good to land this!

File: augly/image/functional.py (hunk 16 above)
Comment (membriux) [reply augly/image/functional.py:928]: Good catch!

File: augly/tests/image_tests/base_unit_test.py [hunk 17]
```diff
@@ -96,13 +96,17 @@ def evaluate_class(
         transform_class: Callable[..., Image.Image],
         fname: str,
         metadata_exclude_keys: Optional[List[str]] = None,
+        check_mode: Optional[bool] = True,
```
Comment (jbitton) [inline augly/tests/image_tests/base_unit_test.py:99]:
```suggestion
        check_mode: bool = True,
```
nit -- the optional type just means that we can pass in type X or None, since we're only ever passing in booleans, this is not needed!

**Comment by membriux** [comment]
Thanks for catching those @jbitton! Ran black on those files as well :)
